File: src/main/java/net/igneo/icv/entity/helmet/blackHole/BlackHoleEntity.java
package net.igneo.icv.entity.helmet.blackHole;

import net.igneo.icv.entity.ICVEntity;
import net.minecraft.world.damagesource.DamageSource;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.projectile.Projectile;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.Vec3;
import software.bernie.geckolib.animatable.GeoEntity;
import software.bernie.geckolib.core.animation.AnimationState;
import software.bernie.geckolib.core.animation.RawAnimation;
import software.bernie.geckolib.core.object.PlayState;

public class BlackHoleEntity extends ICVEntity {
    
    protected static final RawAnimation IDLE_ANIM = RawAnimation.begin().thenLoop("idle");
    
    public BlackHoleEntity(EntityType<? extends Projectile> pEntityType, Level pLevel) {
        super(pEntityType, pLevel);
    }
    
    @Override
    protected <E extends GeoEntity> PlayState animController(AnimationState<E> event) {
        return event.setAndContinue(IDLE_ANIM);
    }
    
    @Override
    public boolean isNoGravity() {
        return true;
    }
    
    @Override
    public void tick() {
        super.tick();
        if (this.tickCount > 400 || this.getOwner() == null) {
            this.discard();
        }
        
        pull();
    }
    
    private void pull() {
        for (Entity entity : this.level().getEntities(null, this.getBoundingBox().inflate(6))) {
            if (entity != this) {
                double scale = 0.05;
                Vec3 pushVec = entity.position().subtract(this.position()).reverse();
                pushVec = pushVec.normalize().scale(scale);
                entity.addDeltaMovement(pushVec);
                if (entity.distanceTo(this) < 2) {
                    entity.addDeltaMovement(pushVec);
                    entity.hurt(this.damageSources().cramming(), 3);
                }
            }
        }
    }
    
    @Override
    public double getAirFriction() {
        return 0.1;
    }
    
    
    @Override
    public boolean isPickable() {
        return true;
    }
    
    @Override
    public boolean hurt(DamageSource pSource, float pAmount) {
        if (pSource.getEntity() == null) {
            Vec3 pushVec = pSource.getSourcePosition().subtract(this.position()).normalize().reverse().scale(1);
            this.addDeltaMovement(new Vec3(
                    pushVec.x,
                    0,
                    pushVec.z
            ));
        } else if (pSource.getSourcePosition() != null) {
            this.addDeltaMovement(pSource.getEntity().getLookAngle().normalize().scale(0.2));
        }
        return super.hurt(pSource, pAmount);
    }
    
    @Override
    protected void onHitBlock(BlockHitResult pResult) {
    
    }
    
}

File: src/main/java/net/igneo/icv/entity/helmet/blackHole/BlackHoleEntity.kt
package net.igneo.icv.entity.helmet.blackHole

import net.igneo.icv.entity.ICVEntity
import net.minecraft.world.damagesource.DamageSource
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.projectile.Projectile
import net.minecraft.world.level.Level
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.Vec3
import software.bernie.geckolib.animatable.GeoEntity
import software.bernie.geckolib.core.animation.AnimationState
import software.bernie.geckolib.core.animation.RawAnimation
import software.bernie.geckolib.core.`object`.PlayState

class BlackHoleEntity(pEntityType: EntityType<out Projectile?>, pLevel: Level) :
    ICVEntity(pEntityType, pLevel) {

    override fun <E : GeoEntity?> animController(event: AnimationState<E>?): PlayState? {
        return event!!.setAndContinue(IDLE_ANIM)
    }

    override fun isNoGravity(): Boolean {
        return true
    }

    override fun tick() {
        super.tick()
        if (this.tickCount > 400 || this.owner == null) {
            this.discard()
        }

        pull()
    }

    private fun pull() {
        for (entity in level().getEntities(null, this.boundingBox.inflate(6.0))) {
            if (entity !== this) {
                val scale = 0.05
                var pushVec = entity.position().subtract(this.position()).reverse()
                pushVec = pushVec.normalize().scale(scale)
                entity.addDeltaMovement(pushVec)
                if (entity.distanceTo(this) < 2) {
                    entity.addDeltaMovement(pushVec)
                    entity.hurt(damageSources().cramming(), 3f)
                }
            }
        }
    }

    override val airFriction: Double
        get() = 0.1


    override fun isPickable(): Boolean {
        return true
    }

    override fun hurt(pSource: DamageSource, pAmount: Float): Boolean {
        if (pSource.entity == null) {
            val pushVec: Vec3 = pSource.sourcePosition!!.subtract(this.position()).normalize().reverse().scale(1.0)
            this.addDeltaMovement(
                Vec3(
                    pushVec.x,
                    0.0,
                    pushVec.z
                )
            )
        } else if (pSource.sourcePosition != null) {
            this.addDeltaMovement(pSource.entity!!.lookAngle.normalize().scale(0.2))
        }
        return super.hurt(pSource, pAmount)
    }

    override fun onHitBlock(pResult: BlockHitResult) {
    }

    companion object {
        protected val IDLE_ANIM: RawAnimation = RawAnimation.begin().thenLoop("idle")
    }
}
